collect_app/src/main/java/org/odk/collect/android/preferences/screens/ProjectPreferencesFragment.kt

```diff
@@ -39,90 +45,167 @@ class ProjectPreferencesFragment :
     override fun onAttach(context: Context) {
         super.onAttach(context)
         DaggerUtils.getComponent(context).inject(this)
+        setHasOptionsMenu(true)
+
+        val changeAdminPasswordViewModel = ViewModelProvider(requireActivity()).get(
+            ChangeAdminPasswordViewModel::class.java
+        )
+        changeAdminPasswordViewModel.passwordSet.observe(
+            this,
+            { isPasswordSet: Boolean ->
+                if (isPasswordSet) {
+                    projectPreferencesViewModel.setStateUnlocked()
```

Comment:
I think we can potentially simplify this. Right now, we have two ViewModels interacting via observations from Fragments. 

Instead, could we share the `ProjectPreferencesViewModel` between the `ProjectPreferencesFragment` and the `AdminPasswordDialogFragment`. I'm imagining the ViewModel could have an `attemptPassword` method that checks the password and then sets the locked/unlocked state based on that. The locked state could then also be a `LiveData` exposed by the `ProjectPreferencesViewModel` and the fragment could decide how it renders based on an observation of that. 

For showing Toasts etc I think exposing a separate `LiveData` for "errors" or "failed attempts" could work. This could use a `LiveData<Consumable>` instead of a `SingleEventLiveData`. I prefer this as it allows observers to make their own decision about what to do with the "event" rather than changing the semantics of `LiveData`. 

Does that make sense @grzesiek2010 @lognaturel?

Reply:
Like this https://github.com/getodk/collect/pull/4687/commits/a2ba0a754b27b5bcaf58cf8ed78ab48478612dab?
Yeah it makes the code a bit cleaner. I forgot about `LiveData<Consumable>` I can use it but I'm still not sure if I like it more than that `SingleEventLiveData`...

Reply:
Yup, liking the new structure a lot more! I'm wondering if we could get rid of the `Consumable` entirely, but I think there is some awkwardness with the prefs lifecycle and `LiveData` that we'd need to get right, so probably a tweak for another day!